src/mozanalysis/experiment.py

```diff
@@ -0,0 +1,508 @@
+        """Return a DataFrame of enrolled clients.
+
+        This works for pref-flip and addon studies.
+
+        The underlying queries are different for pref-flip vs addon
+        studies, because as of 2019/04/02, branch information isn't
+        reliably available in the `events` table for addon experiments:
+        branch may be NULL for all enrollments. The enrollment
+        information for them is most reliably available in
+        `telemetry_shield_study_parquet`. Once this issue is resolved,
+        we will probably start using normandy events for all desktop
+        studies.
+        Ref: https://bugzilla.mozilla.org/show_bug.cgi?id=1536644
+
+        Args:
+            spark: The spark context.
+            study_type (str): One of the following strings:
+                - 'pref_flip'
+                - 'addon'
+            end_date (str, optional): Ignore enrollments after this
+                date: for faster queries on stale experiments. If you
+                set `num_dates_enrollment` then do not set this; at best
+                it would be redundant, at worst it's contradictory.
+
+        Returns:
+            A Spark DataFrame of enrollment data. One row per
+            enrollment. Columns:
+                - client_id (str)
+                - enrollment_date (str): e.g. '20190329'
+                - branch (str)
+        """
+        if study_type == 'pref_flip':
+            enrollments = self._get_enrollments_view_normandy(spark)
+
+        elif study_type == 'addon':
+            enrollments = self._get_enrollments_view_addon(spark, self.addon_version)
+
+        # elif study_type == 'glean':
+        #     raise NotImplementedError
+
+        else:
+            raise ValueError("Unrecognized study_type: {}".format(study_type))
+
+        enrollments = enrollments.filter(
+            enrollments.enrollment_date >= self.start_date
+        ).filter(
+            enrollments.experiment_slug == self.experiment_slug
+        )
+
+        if self.num_dates_enrollment is not None:
+            assert end_date is None, "Possibly contradictory instructions!"
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= self._get_scheduled_max_enrollment_date()
+            )
+        elif end_date is not None:
+            enrollments = enrollments.filter(
+                enrollments.enrollment_date <= end_date
+            )
+
+        enrollments.cache()
+
+        return enrollments
+
+    def get_per_client_data(
+        self, enrollments, df, metric_list, last_date_full_data, conv_window_start_days,
+        conv_window_length_days, keep_client_id=False
+    ):
+        """Return a DataFrame containing per-client metric values.
+
+        Args:
+            enrollments: A spark DataFrame of enrollments, like the one
+                returned by `self.get_enrollments()`.
+            df: A spark DataFrame containing the data needed to calculate
```

Comment:
I'd love a more informative public name for this than `df`. `ping_data`? `metric_source`? `measurements`?

Reply:
I'd love a better name too, but this is trying to be a very generic quantity: as per the docs it is "A spark DataFrame containing the data needed to calculate the metrics". This data won't always be "ping data" or "measurements" (e.g. I don't think crash reports fall into either category). Is there a terse name that conveys "data that is going to be turned into a metric"? I think `metric_source` suggests that the data is already metricified (or emetricuted?) but I want to convey that it's OK to chuck in generic tables?

Reply:
Yeah I was reaching for suggestions. I probably assert a row of crash_summary reflects a measurement of a crash, and it's definitely a summary of a ping; &shrug;!